test/python/circuit/test_gate_power.py

```diff
@@ -0,0 +1,257 @@
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+
+"""Test Qiskit's power instruction operation."""
+
+import unittest
+from ddt import ddt, data
+from numpy import array, eye
+from numpy.testing import assert_allclose, assert_array_almost_equal
+from numpy.linalg import matrix_power
+
+from qiskit.transpiler import PassManager
+from qiskit import QuantumRegister, QuantumCircuit
+from qiskit.test import QiskitTestCase
+from qiskit.extensions import SGate, UnitaryGate, CnotGate
+from qiskit.circuit import Gate
+from qiskit.transpiler.passes import Unroller, Optimize1qGates
+from qiskit.quantum_info.operators import Operator
+
+
+@ddt
+class TestPowerSgate(QiskitTestCase):
+    """Test Sgate.power() with integer"""
+
+    @data(2, 1, 0, -1, -2)
+    def test_sgate_int(self, n):
+        """Test Sgate.power(n) method with n as integer.
+        """
+        result = SGate().power(n)
+
+        self.assertEqual(result.label, 's^%s' % n)
+        self.assertIsInstance(result, UnitaryGate)
+        assert_array_almost_equal(result.to_matrix(), matrix_power(SGate().to_matrix(), n))
+
+    def test_sgate_float_1_5(self, n=1.5):
+        """Test Sgate.power(1.5) method.
+        """
+        result = SGate().power(n)
+
+        expected = array([[1, 0],
+                          [0, -0.70710678 + 0.70710678j]], dtype=complex)
+        self.assertEqual(result.label, 's^%s' % n)
+        self.assertIsInstance(result, UnitaryGate)
+        assert_array_almost_equal(result.to_matrix(), expected)
+
+    def test_sgate_float_0_1(self, n=0.1):
+        """Test Sgate.power(0.1) method.
+        """
+        result = SGate().power(n)
+
+        expected = array([[1, 0],
+                          [0, 0.98768834 + 0.15643447j]], dtype=complex)
+        self.assertEqual(result.label, 's^%s' % n)
+        self.assertIsInstance(result, UnitaryGate)
+        assert_array_almost_equal(result.to_matrix(), expected)
+
+    def test_sgate_float_minus_1_5(self, n=-1.5):
+        """Test Sgate.power(-1.5) method.
+        """
+        result = SGate().power(n)
+
+        expected = array([[1, 0],
+                          [0, -0.70710678 - 0.70710678j]], dtype=complex)
+        self.assertEqual(result.label, 's^%s' % n)
+        self.assertIsInstance(result, UnitaryGate)
+        assert_array_almost_equal(result.to_matrix(), expected)
+
+    def test_sgate_float_minus_0_1(self, n=-0.1):
+        """Test Sgate.power(-0.1) method.
+        """
+        result = SGate().power(n)
+
+        expected = array([[1, 0],
+                          [0, 0.98768834 - 0.15643447j]], dtype=complex)
+        self.assertEqual(result.label, 's^%s' % n)
+        self.assertIsInstance(result, UnitaryGate)
+        assert_array_almost_equal(result.to_matrix(), expected)
```

Comment:
There is a lot of repetitions in these tests. Can you use `ddt` for all of them? 250 lines of test is a bit extreme for testing a simple feature like this and it will make it hard to parse through these tests later. I understand that the ground truth matrix has been hardcoded in these and that's why each one is a different test case, but I think you can just use `matrix_power` to compute them like you have done for the other `ddt` cases. Presumably that's what you have done anyway to get the ground truths.

Reply:
I moved to ddt in acb5232a. `matrix_power` only works with integers. Have a look.

Reply:
Similarly, d0807be5.

Reply:
currently 183 lines...